README.md
# BotMan Web Widget

This is the Sunday IT web widget. Forked from [botman/web-widget](https://github.com/botman/web-widget) but heavily modified. 
This widget can be used with default Botman 2.0 installations, but best used with [molbal/sundayit-chatbot-admin](https://github.com/molbal/sundayit-chatbot-admin)

With the extended control panel functionality, this widget sends some metadata and can reecive messages back from the control panel.

# Installation

## 1. Clone this repository
```$ git clone https://github.com/sunday-ltd/web-widget```

## Install dependencies
```$ npm install```

## Run development environment
```$ npm run dev```


#### Widget
Next you need to add the widget to your website and tell it to use the chat frame URL that you just created.
Simply add the Javascript to your website and configure it using a global config object. Place the URL that you created for the chat frame in the `frameEndpoint` configuration key.


```html
<script>
var botmanWidget = {
        chatServer: 'http://127.0.0.1:8000,
        frameEndpoint: 'chat.html',
        pingLocation: "http://127.0.0.1:8000/chat-activity"
};
</script>
<script src='https://cdn.jsdelivr.net/npm/botman-web-widget@0/build/js/widget.js'></script>
```
##### All configuration options
When adding the widget to your own website, you probably want to customize the look of the widget, the title that get's displayed in the widget bar, background images etc. That's why the web widget comes with a detailed set of configuration options.

To apply these options to your widget, just declare an option called botmanWidget before loading the widget's Javascript.

The UI of the chat widget itself can simply be customized via CSS - take a look at the default CSS to see the available selectors.

| Key                      | Default                   | Description                                                                                                     |
|--------------------------|---------------------------|-----------------------------------------------------------------------------------------------------------------|
| `chatServer`             | `/botman`                 | The URL of the BotMan route / server to use.                                                                    |
| `frameEndpoint`          | `chat.html`               | The location of your chat frame URL / route.                                                                    |
| `timeFormat`             | `HH:MM`                   | Time format to use                                                                                              |
| `dateTimeFormat`         | `m/d/yy HH:MM`            | Date-Time format to use                                                                                         |
| `title`                  | `Chat`                    | The title to use in the widget.                                                                                 |
| `introMessage`           | `null`                    | This is a welcome message that every new user sees when the widget is opened for the first time.                |
| `placeholderText`        | `Send a message...`       | Input placeholder text                                                                                          |
| `displayMessageTime`     | `true`                    | Determine if message times should be shown                                                                      |
| `mainColor`              | `#f37820`                 | The main color used in the widget header.                                                                       |
| `headerTextColor`        | `#fff`                    | Color for the header  text                                                                                      |
| `bubbleAvatarUrl`        | `null`                        | The image url to use in the chat bubble.                                                                        |
| `desktopHeight`          | `450`                     | Height of the opened chat widget on desktops.                                                                   |
| `desktopWidth`           | `370`                     | Width of the opened chat widget on desktops.                                                                    |
| `mobileHeight`           | `100%`                    | Height of the opened chat widget on mobile.                                                                     |
| `mobileWidth`            | `300`                     | Width of the opened chat widget on mobile.                                                                      |
| `videoHeight`            | `160`                     | Height to use for embedded videos.                                                                              |
| `userId`                 |                           | Optional user-id that get's sent to BotMan. If no ID is given, a random id will be generated on each page-view. |
| `headerTextFontFamily`   | `'Open Sans', sans-serif` | Header text font family                                                                                         |
| `headerTextFontWeight`   | `bolder`                  | Header text font weight                                                                                         |
| `headerIconBorderRadius` | `0,5`                     | Border radius for the icon at the top                                                                           |
| `headerIconDisplayed`    | `true`                    | Should draw header icon?                                                                                        |
| `headerIconSize`         | `72px`                    | Size for the header icon (width and height)                                                                     |
| `pingLocation`           | `false`                   | Sunday admin service's ping location                                                                            |

## API functions

The web widget also comes with an API to programatically send messages from the user or chatbot. You might use this feature when the user clicks on a button on your website to instantly trigger the chatbot, without having the user to type something. Once the chat widget is initialized, you can access it's API on the botmanChatWidget object that get's exposed to the window containing the chat widget.

|   Method                          | Description                                                                   |
|-----------------------------------|-------------------------------------------------------------------------------|
| `botmanChatWidget.open()`         | Open the chat widget.                                                         |
| `botmanChatWidget.close()`        | Close the chat widget.                                                        |
| `botmanChatWidget.toggle()`       | Toggle the chat widget.                                                       |
| `botmanChatWidget.say(text)`      | Say something on behalf of the user. The given text is visible in the widget. |
| `botmanChatWidget.whisper(text)`  | Similar to say, with the difference that the text is not visible.             |
| `botmanChatWidget.sayAsBot(text)` | Say something on behalf of the chatbot. The text is visible in the widget.    |
| `botmanChatWidget.isOpen()`       | Returns whether the widget is open or not.                                    |

# Build, add to Laravel

See the chatframe repository, [here](https://github.com/sunday-ltd/chatframe)

# Changelog

See Changelog.md


## Credits
![](https://img.icons8.com/cotton/64/000000/cloud-user-group.png)

- [Sunday Ltd.](https://sundayit.hu/chatbots)
- [Botman](https://botman.io) / [Botman docs](https://github.com/botman/docs/blob/master/web-widget.md)
- [Icons8](https://icons8.com)


Changelog.md
![](https://sundayit.hu/img/logo.png) 
# Changelog

![](https://img.icons8.com/cotton/64/000000/document.png)

All notable changes to `chatframe` will be documented in this file.

## Version 1.2.0.2
### Changes
- Bugfix release, fixing a broken integrity hash and a history retrieval bug which sent back the retreived data again to the server.

## Version 1.2.0.1
### Changes
- Adds retrieval of administrator messages through the admin endpoint from [Sunday Chatbot Control](https://github.com/molbal/sundayit-chatbot-admin)
- Reloads messages after reloading the site from the [Sunday Chatbot Control](https://github.com/molbal/sundayit-chatbot-admin) server side through the admin endpoint

## Version 1.1
### Changes
- Adds functionality to ping the backend for [Sunday Chatbot Control](https://github.com/molbal/sundayit-chatbot-admin)
- Adds better documentation of configuration options and API methods.

## Version 1.0

### Added
- Initial release after fork from botman/web-widget


src/widget/api.ts
import Widget from './widget';
import {IMessage} from "../typings";

export default class Api {

    widget: Widget;

    constructor(widget: Widget) {
        this.widget = widget;
    }

    open() {

        this.widget.open();
    }

    close() {
        this.widget.close();
    }

    toggle() {
        this.widget.toggle();
    }

    isOpen() {
        return this.widget.state.isChatOpen === true;
    }

    callChatWidget(payload: Object) {
        console.log("sending message to chat iframe: " , payload);
        if (this.isOpen()) {
            (document.getElementById('chatBotManFrame') as HTMLIFrameElement).contentWindow.postMessage(payload, '*');
        } else {
            try {
                this.open();
                setTimeout(() => {
                    (document.getElementById('chatBotManFrame') as HTMLIFrameElement).contentWindow.postMessage(payload, '*');
                }, 750);
            } catch (e) {
                console.error(e);
            }
        }
    }

    writeToMessages(message: IMessage) {
        this.callChatWidget({
            method: 'writeToMessages',
            params: [
                message
            ]
        })
    }

    sayAsBot(text: string) {
        this.callChatWidget({
            method: 'sayAsBot',
            params: [
                text
            ]
        });
    }

    say(text: string) {
        this.callChatWidget({
            method: 'say',
            params: [
                text
            ]
        });
    }

    whisper(text: string) {
        this.callChatWidget({
            method: 'whisper',
            params: [
                text
            ]
        });
    }

}


src/chat/botman.ts
import axios from 'axios';
import { IAttachment, IMessage } from './../typings';

class BotMan {

	userId: string;
	chatServer: string;

    setUserId(userId: string) {
        this.userId = userId;
    }

    setChatServer(chatServer: string) {
        this.chatServer = chatServer;
    }

    callAPI = (text: string, interactive = false, attachment: IAttachment = null, perMessageCallback: Function, callback: Function) => {
    	let data = new FormData();
    	const postData: { [index: string] : string|Blob } = {
    		driver: 'web',
    		userId: this.userId,
    		message: text,
    		attachment: attachment as Blob,
    		interactive: interactive ? '1' : '0'
    	};

    	Object.keys(postData).forEach(key => data.append(key, postData[key]));

    	axios.post(this.chatServer, data).then(response => {
    		const messages = response.data.messages || [];

			if (perMessageCallback) {
				messages.forEach((msg: IMessage) => {
					perMessageCallback(msg);
				});
			}

    		if (callback) {
    			callback(response.data);
    		}
    	});
    };

    /**
     *
     * @param url
     * @param callBack
     */
    callPing = (url: string, callBack: Function) => {
        let data = new FormData();

        const postData = {
            userId: this.userId,
            widgetOpened: window.botmanChatWidget.isOpen()
        };
    };

}

export let botman = new BotMan();


src/widget/style.ts

export const desktopWrapperStyle = {
    position: 'fixed',
    bottom: '20px',
    right: '20px',
    zIndex: 2147483647,
    borderRadius: '5px',
    boxSizing: 'content-box',
    boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.2)',
    // overflow: 'hidden'
};

export const desktopClosedWrapperStyleChat = {
    position: 'fixed',
    bottom: '0px',
    right: '0px',
    zIndex: 2147483647,
    minWidth: '400px',
    boxSizing: 'content-box',
    overflow: 'hidden',
    minHeight: '120px'
};

export const mobileClosedWrapperStyle = {
    position: 'fixed',
    bottom: '0px',
    right: '0px',
    zIndex: 2147483647,
    minWidth: '400px',
    boxSizing: 'content-box',
    overflow: 'hidden',
    minHeight: '120px'
};

export const mobileOpenWrapperStyle = {
    position: 'fixed',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    zIndex: 2147483647,
    width: '100%',
    height: '100%',
    overflowY: 'visible',
    boxSizing: 'content-box'
};

export const desktopTitleStyle = {
    height: '40px',
    lineHeight: '30px',
    fontSize: '20px',
    display: 'flex',
    justifyContent: 'space-between',
    padding: '5px 0 5px 20px',
    fontFamily: 'Lato, sans-serif',
    color: '#fff',
    cursor: 'pointer',
    boxSizing: 'content-box',
    mozBoxSizing: 'content-box',
    webkitBoxSizing: 'content-box',
    borderRadius: '10px 10px 0 0'
};

export const mobileTitleStyle = {
    height: 52,
    width: 52,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    webkitBoxShadow: '0px 0px 20px rgba(0, 0, 0, 0.2)',
    boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.2)',
    cursor: 'pointer',
    borderRadius: '50%',
};

export const desktopClosedMessageStyle = {
    letterSpacing: '1px',
    color: '#fff',
    display: 'block',
    position: 'absolute',
    fontSize: '14px',
    top: 0,
    right: '130px',
    marginTop: '23px',
    borderRadius: '5px',
    padding: '15px 20px',
    boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.2)'
};

export const desktopIntroMessageStyle = {
    width: 0,
    height: 0,
    position: 'absolute',
    top: '12px',
    right: '-10px',
    borderTop: '10px solid transparent',
    borderBottom: '10px solid transparent'
};

export const desktopClosedMessageAvatarStyle = {
    display: 'flex',
    justifyContent: 'center',
    position: 'absolute',
    top: '38px',
    right: '20px',
    height: '60px',
    width: '60px',
    border: 0,
    borderRadius: '50%',
    boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.2)'
};

export const mobileClosedMessageAvatarStyle = {
    display: 'block',
    position: 'absolute',
    top: '46px',
    right: '20px',
    height: '52px',
    width: '52px',
    border: 0,
    borderRadius: '50%',
    boxShadow: '0px 0px 20px rgba(0, 0, 0, 0.2)'
};

export const closedChatAvatarImageStyle = {
    width: '100%',
    height: 'auto',
    borderRadius: '999px'
};


src/typings/index.ts


export interface IMessageTypeProps {
    message: IMessage,
    calculatedTimeout?: number,
    messageHandler: Function,
    onVisibilityChange: Function,
    timeout: number,
    conf?: IConfiguration,
}

export interface IMessageHolderProps {
    message: IMessage,
    calculatedTimeout: number,
    messageHandler: Function,
    conf?: IConfiguration,
}

export interface IMessageTypeState {
    visible: boolean,
    visibilityChanged: boolean,
    attachmentsVisible: boolean
}

export interface IMessage {
    id?: string,
    text: string,
    type: "text" | "list",
    timeout?: number,
    from: string,
    time?: string,
    visibilityChanged?: boolean;
    visible?: boolean,
    actions?: IAction[],
    buttons?: IButton[],
    attachment?: IAttachment,
    globalButtons?: IButton[],
    additionalParameters?: any,
    elements?: IElement[],
}

export interface IAttachment {
    url?: string,
    type?: string,
}

export interface IElement {
    title : string,
    image_url: string,
    item_url: string,
    subtitle: string,
    default_action: IAction,
    buttons: IButton[],
}


export interface IButton {
    type: string,
    title: string,
    url: string,
    payload: string,
}

export interface IAction {
    text: string,
    value: string,
}

export interface IConfiguration {
    /**
     * The URL of the BotMan route / server to use.
     */
    chatServer: string,
    /**
     * The location of your chat frame URL / route.
     */
    frameEndpoint: string,
    /**
     * Date and time format to use
     */
    timeFormat: string;
    dateTimeFormat: string,
    /**
     * The title to use in the widget header.
     */
    title: string,
    /**
     * How many days is the cookie valid?
     */
    cookieValidInDays: number,
    /*
     * This is a welcome message that every new user sees
     * when the widget is opened for the first time.
     */
    introMessage: string,
    /*
     * Input placeholder text
     */
    placeholderText: string,
    /*
     * Toggle display of message times
     */
    displayMessageTime: boolean,
    /*
     * Send a widget opened event
     */
    sendWidgetOpenedEvent: boolean,
    widgetOpenedEventData: string,
    mainColor: string,
    headerTextColor: string,
    bubbleBackground: string,
    bubbleAvatarUrl: string,
    desktopHeight: number,
    desktopWidth: number,
    mobileHeight: string,
    mobileWidth: string,
    videoHeight: number,
    aboutLink: string,
    aboutText: string,
    chatId: string,
    userId: string,
    wrapperHeight?: number,
    alwaysUseFloatingButton: boolean,

    useEcho: boolean,

    echoChannel: string | any,

    echoChannelType: string,

    echoConfiguration: any,

    echoEventName: string,

    init?: Function,

    /** SundayIT added functionality */

    /**
     * Header text font family
     */
    headerTextFontFamily: string,

    /**
     * Header text font weight
     */
    headerTextFontWeight: string | number,

    /**
     * Border radius for the icon at the top
     */
    headerIconBorderRadius: string

    /**
     * Size for the header icon (width and height)
     */
    headerIconSize: string

    /**
     * Should draw header icon?
     */
    headerIconDisplayed: boolean,

    /**
     * Sunday's ping location
     */
    pingLocation: string | false
}


webpack.config.babel.js
const webpack = require('webpack');
const ExtractTextPlugin = require('extract-text-webpack-plugin');
const autoprefixer = require('autoprefixer');
const CopyWebpackPlugin = require('copy-webpack-plugin');
const path = require('path');
const ENV = process.env.NODE_ENV || 'development';

const CSS_MAPS = ENV!=='production';

module.exports = {
    context: path.resolve(__dirname, "src"),
    entry: {
        widget: './widget/index.tsx',
        chat: './chat/index.tsx'
    },

    output: {
        path: path.resolve(__dirname, "build/js"),
        publicPath: '/',
        filename: '[name].js'
    },

    resolve: {
        extensions: ['.tsx', '.ts', '.jsx', '.js', '.json', '.less'],
        modules: [
            path.resolve(__dirname, "src/lib"),
            path.resolve(__dirname, "node_modules"),
            'node_modules'
        ],
        alias: {
            typings: path.resolve(__dirname, "src/typings"),
            components: path.resolve(__dirname, "src/components"),    // used for tests
            style: path.resolve(__dirname, "src/style"),
            'react': 'preact-compat',
            'react-dom': 'preact-compat'
        }
    },

    module: {
        rules: [
            {
                test: /\.jsx?$/,
                exclude: path.resolve(__dirname, 'src'),
                enforce: 'pre',
                use: 'source-map-loader'
            },
            { test: /\.tsx?$/, use: { loader: 'awesome-typescript-loader' } },
            {
                // Transform our own .(less|css) files with PostCSS and CSS-modules
                test: /\.(less|css)$/,
                include: [path.resolve(__dirname, 'src/components')],
                use: ExtractTextPlugin.extract({
                    fallback: 'style-loader',
                    use: [
                        {
                            loader: 'css-loader',
                            options: { modules: true, sourceMap: CSS_MAPS, importLoaders: 1, minimize: true }
                        },
                        {
                            loader: `postcss-loader`,
                            options: {
                                sourceMap: CSS_MAPS,
                                plugins: () => {
                                    autoprefixer({ browsers: [ 'last 2 versions' ] });
                                }
                            }
                        },
                        {
                            loader: 'less-loader',
                            options: { sourceMap: CSS_MAPS }
                        }
                    ]
                })
            },
            {
                test: /\.(less|css)$/,
                exclude: [path.resolve(__dirname, 'src/components')],
                use: ExtractTextPlugin.extract({
                    fallback: 'style-loader',
                    use: [
                        {
                            loader: 'css-loader',
                            options: { sourceMap: CSS_MAPS, importLoaders: 1, minimize: true }
                        },
                        {
                            loader: `postcss-loader`,
                            options: {
                                sourceMap: CSS_MAPS,
                                plugins: () => {
                                    autoprefixer({ browsers: [ 'last 2 versions' ] });
                                }
                            }
                        },
                        {
                            loader: 'less-loader',
                            options: { sourceMap: CSS_MAPS }
                        }
                    ]
                })
            },
            {
                test: /\.json$/,
                use: 'json-loader'
            },
            {
                test: /\.(xml|html|txt|md)$/,
                use: 'raw-loader'
            },
            {
                test: /\.(svg|woff2?|ttf|eot|jpe?g|png|gif)(\?.*)?$/i,
                use: ENV==='production' ? 'file-loader' : 'url-loader'
            }
        ]
    },
    plugins: ([
        new webpack.NoEmitOnErrorsPlugin(),
        new webpack.DefinePlugin({
            'process.env.NODE_ENV': JSON.stringify(ENV)
        }),
        new CopyWebpackPlugin([
            { from: './chat.html', to: './../' },
            { from: './demo.html', to: './../' }
        ])
    ]).concat(ENV==='production' ? [
        new webpack.optimize.UglifyJsPlugin({
            output: {
                comments: false
            },
            compress: {
                unsafe_comps: true,
                properties: true,
                keep_fargs: false,
                pure_getters: true,
                collapse_vars: true,
                unsafe: true,
                warnings: false,
                screw_ie8: true,
                sequences: true,
                dead_code: true,
                drop_debugger: true,
                comparisons: true,
                conditionals: true,
                evaluate: true,
                booleans: true,
                loops: true,
                unused: true,
                hoist_funs: true,
                if_return: true,
                join_vars: true,
                cascade: true,
                drop_console: true
            }
        })
    ] : []),

    stats: { colors: true },

    node: {
        global: true,
        process: false,
        Buffer: false,
        __filename: false,
        __dirname: false,
        setImmediate: false
    },

    devtool: ENV==='production' ? 'source-map' : 'cheap-module-eval-source-map',

    devServer: {
        port: process.env.PORT || 8080,
        host: 'localhost',
        publicPath: '/js/',
        contentBase: './src',
        historyApiFallback: true,
        open: true,
        openPage: 'demo.html'
    }
};


src/widget/configuration.ts
import { IConfiguration } from "../typings";

export const defaultConfiguration: IConfiguration = {
    chatServer: '/botman',
    frameEndpoint: 'chat.html',
    timeFormat: 'HH:MM',
    dateTimeFormat: 'm/d/yy HH:MM',
    title: 'Chat',
    cookieValidInDays: 1,
    introMessage: '',
    placeholderText: 'Send a message...',
    displayMessageTime: true,
    sendWidgetOpenedEvent: false,
    widgetOpenedEventData: '',
    mainColor: '#f37820',
    headerTextColor: '#fff',
    bubbleBackground: '#408591',
    bubbleAvatarUrl: '',
    desktopHeight: 450,
    desktopWidth: 370,
    mobileHeight: '100%',
    mobileWidth: '300px',
    videoHeight: 160,
    aboutLink: '',
    aboutText: '',
    chatId: '',
    userId: '',
    alwaysUseFloatingButton: false,
    useEcho: false,
    echoChannel: (userId: string) => '',
    echoConfiguration: {},
    echoEventName: '.message.created',
    echoChannelType: 'private',
    headerTextFontFamily: "'Open Sans', sans-serif",
    headerTextFontWeight: 'bolder',
    headerIconBorderRadius: '50%',
    headerIconDisplayed: true,
    headerIconSize: '72px',
    pingLocation: false
};
